File: library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/IndexTest.java
package com.raizlabs.android.dbflow.test.sql;

import com.raizlabs.android.dbflow.sql.builder.Condition;
import com.raizlabs.android.dbflow.sql.index.Index;
import com.raizlabs.android.dbflow.sql.language.Delete;
import com.raizlabs.android.dbflow.sql.language.Select;
import com.raizlabs.android.dbflow.test.FlowTestCase;

import java.util.List;

/**
 * Description:
 */
public class IndexTest extends FlowTestCase {

    public void testIndex() {

        Delete.table(IndexModel.class);

        Index<IndexModel> modelIndex = new Index<IndexModel>("salary_index")
                .on(IndexModel.class, IndexModel$Table.SALARY);
        modelIndex.disable();

        assertEquals("CREATE INDEX IF NOT EXISTS `salary_index` ON `IndexModel`(`salary`)", modelIndex.getQuery().trim());

        modelIndex.enable();

        IndexModel indexModel = new IndexModel();
        indexModel.name = "Index";
        indexModel.salary = 30000;
        indexModel.save();

        indexModel = new IndexModel();
        indexModel.name = "Index2";
        indexModel.salary = 15000;
        indexModel.save();

        List<IndexModel> list = new Select().from(IndexModel.class)
                .indexedBy(modelIndex.getIndexName())
                .where(Condition.column(IndexModel$Table.SALARY).greaterThan(20000)).queryList();

        assertTrue(list.size() == 1);

        modelIndex.disable();

        Delete.table(IndexModel.class);

    }
}

File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/sql/language/IndexTest.kt
package com.raizlabs.android.dbflow.sql.language

import com.raizlabs.android.dbflow.BaseUnitTest
import com.raizlabs.android.dbflow.config.databaseForTable
import com.raizlabs.android.dbflow.models.SimpleModel
import com.raizlabs.android.dbflow.models.SimpleModel_Table
import org.junit.Assert.assertEquals
import org.junit.Test

class IndexTest : BaseUnitTest() {

    @Test
    fun validateBasicIndex() {
        databaseForTable<SimpleModel> {
            assertEquals("CREATE INDEX IF NOT EXISTS `index` ON `SimpleModel`(`name`)",
                    indexOn<SimpleModel>("index", SimpleModel_Table.name).query)
        }
    }

    @Test
    fun validateUniqueIndex() {
        databaseForTable<SimpleModel> {
            assertEquals("CREATE UNIQUE INDEX IF NOT EXISTS `index` ON `SimpleModel`(`name`, `test`)",
                    indexOn<SimpleModel>("index").unique(true).and(SimpleModel_Table.name)
                            .and("test".nameAlias).query)
        }
    }

    @Test
    fun validateBasicIndexNameAlias() {
        databaseForTable<SimpleModel> {
            assertEquals("CREATE INDEX IF NOT EXISTS `index` ON `SimpleModel`(`name`, `test`)",
                    indexOn<SimpleModel>("index", "name".nameAlias, "test".nameAlias).query)
        }
    }
}